### PR TITLE
Integrate group-by plan optimizations into LogicalPlan operator

### PR DESCRIPTION
**Pull request by mfussenegger**
Split off a part of https://github.com/crate/crate/pull/6334

### CODE REVIEW

File: sql/src/main/java/io/crate/planner/node/dql/Collect.java [hunk 1]
```diff
@@ -147,6 +148,19 @@ public int numOutputs() {
     }
 
     @Override
+    public boolean executesOnShard() {
+        List<Projection> projections = collectPhase.projections();
+        if (projections.isEmpty()) {
+            return collectPhase instanceof RoutedCollectPhase &&
```
Comment (matriv) [inline sql/src/main/java/io/crate/planner/node/dql/Collect.java]: Please add a comment for what is done here.

Comment (matriv) [inline sql/src/main/java/io/crate/planner/node/dql/Collect.java]: Is it better to add a property to the CollectPhase and avoid instanceof?

File: sql/src/main/java/io/crate/planner/operators/Collect.java [hunk 2]
```diff
@@ -202,8 +208,8 @@ public LogicalPlan tryCollapse() {
     }
 
     @Override
-    public RowGranularity dataGranularity() {
-        return tableInfo.rowGranularity();
+    public boolean preferShardProjections() {
+        return tableInfo instanceof DocTableInfo;
```
Comment (matriv) [inline sql/src/main/java/io/crate/planner/operators/Collect.java]: Why is that necessary?

File: sql/src/main/java/io/crate/planner/node/dql/Collect.java (hunk 1 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/planner/node/dql/Collect.java]: I don't think it matters - instanceof is not *that* expensive.

File: sql/src/main/java/io/crate/planner/operators/GroupHashAggregate.java [hunk 3]
```diff
@@ -82,58 +85,89 @@ public Plan build(Planner.Context plannerContext,
...
+            ExecutionPhases.executesOnHandler(plannerContext.handlerNode(), plan.resultDescription().nodeIds())) {
+
             GroupProjection groupProjection = projectionBuilder.groupProjection(
-                source.outputs(),
+                sourceOutputs,
                 groupKeys,
                 aggregates,
                 AggregateMode.ITER_FINAL,
-                RowGranularity.CLUSTER
+                // if not clusteredByGroup by we need to use CLUSTER/NODE granularity to merge across shards
```
Comment (matriv) [inline sql/src/main/java/io/crate/planner/operators/GroupHashAggregate.java]: `clusteredByGroup by` => `clusteredByGroupBy`
